test/functional/Valhalla/src/org/openj9/test/lworld/ValueTypeArrayTests.java

```diff
@@ -0,0 +1,358 @@
+				+staticArrayKind+"; actualSourceKind == "+actualSourceKind+"; staticSourceKind == "+staticSourceKind);
+		}
+
+		if (expectedExceptionClass != null) {
+			assertTrue(caughtThrowable,
+				"Expected exception kind "+expectedExceptionClass.getName()+" to be thrown.  ActualArrayKind == "+actualArrayKind
+				+"; StaticArrayKind == "+staticArrayKind+"; actualSourceKind == "+actualSourceKind+"; staticSourceKind == "
+				+staticSourceKind);
+		}
+
+		// ArrayIndexOutOfBoundsException must be checked before both
+		// NullPointerException for primitive value type and ArrayStoreException.
+		// This call to assignDispatch will attempt an out-of-bounds element assignment,
+		// and is always expected to throw an ArrayIndexOutOfBoundsException.
+		boolean caughtAIOOBE = false;
+
+		try {
+			assignDispatch(arr, 2, sourceVal, staticArrayKind, staticSourceKind);
+		} catch(ArrayIndexOutOfBoundsException aioobe) {
+			caughtAIOOBE = true;
+		} catch (Throwable t) {
+			assertTrue(false, "Expected ArrayIndexOutOfBoundsException, but saw "+t.getClass().getName());
+		}
+
+		assertTrue(caughtAIOOBE, "Expected ArrayIndexOutOfBoundsException");
+	}
+
+	/**
+	 * Test various types of arrays and source values along with various statically declared types
+	 * for the arrays and source values to ensure required <code>ArrayStoreExceptions</code>,
+	 * <code>ArrayIndexOutOfBoundsExceptions</code> and <code>NullPointerExceptions</code> are
+	 * detected.
+	 * @throws java.lang.Throwable if the attempted array element assignment throws an exception
+	 */
+	@Test(priority=1,invocationCount=2)
+	static public void testValueTypeArrayAssignments() throws Throwable {
+		Object[][] testArrays = new Object[][] {new Object[2], new SomeIface[2], new PointV[2], new PointPV[2]};
+		int[] kinds = {OBJ_TYPE, IFACE_TYPE, VAL_TYPE, PRIM_TYPE};
+		Object[] vals = new Object[] {null, bogusIfaceObj, new PointV(1.0, 2.0), new PointPV(3.0, 4.0)};
+
+		for (int i = 0; i < testArrays.length; i++) {
+			Object[] testArray = testArrays[i];
+
+			for (int j = 0; j < kinds.length; j++) {
+				int staticArrayKind = kinds[j];
+
+				for (int k = 0; k < kinds.length; k++) {
+					int staticValueKind = kinds[k];
+					// runTest's parameters are of type Object[] and Object.  It ultimately dispatches to an assign
+					// method whose parameters have more specific static types.  This condition filters out the
+					// combinations of static types that aren't permitted from the point of view of the Java language.
+					//
+					// Cases:
+					//  - the two types are the same  (staticArrayKind == staticValueKind)
+					//  OR
+					//    - the type of the array is Object[] or SomeIface[]  (staticArrayKind < VAL_TYPE)
+					//      AND
+					//    - the type of the array is less specific than that of the source (staticArrayKind < staticValueKind)
+					//
+					if (staticArrayKind == staticValueKind || staticArrayKind < staticValueKind && staticArrayKind < VAL_TYPE) {
```

Comment:
Please put `staticArrayKind < staticValueKind && staticArrayKind < VAL_TYPE` inside `( )`

Reply:
Force pushed this change in commit 070b9774fd703021dcdbec2728e5bcf7d53e307a